### PR TITLE
adding a FromSourceAsset StepInputSource

### PR DESCRIPTION
**Pull request by OwenKephart**
### Summary & Motivation

Right now, the way we load assets when we don't have an upstream output connected to an op's input is to use a root_input_manager. This requires that we set a root_manager_key on all InputDefinitions that we want to potentially load this way.

This works ok with things that we decorate with @asset, as we can construct an op that conforms to this requirement, but for graph-backed assets this is not something we can reliably do.

This PR adds a new StepInputSource type to represent scenarios where we'd like to load an input from a source asset

### How I Tested These Changes

doesn't work yet!

### CODE REVIEW

**Comment by OwenKephart** [comment]
Ok so this is still a bit messy, but everything is now wired through properly.

I replicated the behavior we have currently where you're allowed to refer to the `input_context.upstream_output` value even though there is no actual upstream output when loading from a source asset.

This is basically a convenience feature to enable a pattern that people often employ in IOManagers where you have a single, let's say `_get_path()` function, and on the output side you call `self._get_path(context)` and on the input side you call `self._get_path(context.upstream_output)`.

To be honest, this seems like a bit of an anti-pattern if you're building an asset io manager (your load_input function should generally just be a function of asset_key, which is available on either context directly), but I decided to bias towards compatibility for this PR.

File: python_modules/dagster/dagster/core/asset_defs/source_asset.py [hunk 1]
```diff
@@ -24,8 +24,7 @@ class SourceAsset(
         ],
     )
 ):
-    """A SourceAsset represents an asset that is not generated by any Dagster op in the repository
-    that it's referenced from.
+    """A SourceAsset represents an asset that will be loaded by, but not updated by, a given job.
```
Comment (sryza) [inline python_modules/dagster/dagster/core/asset_defs/source_asset.py:27]: I think source assets are a valid concept even in a world without jobs. Is there a way to document this in a way that communicates that?

File: python_modules/dagster/dagster/core/definitions/asset_layer.py [hunk 2]
```diff
@@ -415,10 +438,14 @@ def from_graph_and_assets_node_mapping(
         asset_key_by_input: Dict[NodeInputHandle, AssetKey] = {}
         asset_info_by_output: Dict[NodeOutputHandle, AssetOutputInfo] = {}
         asset_deps: Dict[AssetKey, AbstractSet[AssetKey]] = {}
+        asset_io_managers: Dict[AssetKey, str] = {
```
Comment (sryza) [inline python_modules/dagster/dagster/core/definitions/asset_layer.py:441]: extreme nitpick: naming this `io_manager_keys_by_asset_key` would make it more clear  where it's going to be used

File: python_modules/dagster/dagster/core/execution/context/system.py [hunk 3]
```diff
@@ -493,14 +493,19 @@ def for_input_manager(
         source_handle: Optional[StepOutputHandle] = None,
         resource_config: Any = None,
         resources: Optional["Resources"] = None,
+        artificial_output_context: Optional["OutputContext"] = None,
```
Comment (sryza) [inline python_modules/dagster/dagster/core/execution/context/system.py:496]: Not a change I'm suggesting making in this PR, but I wonder if we should stop including an artificial output context in these situations.

**Review by sryza — COMMENTED** [review]
Looking good! Still a draft?

File: python_modules/dagster/dagster/core/asset_defs/source_asset.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/core/asset_defs/source_asset.py:27]: maybe i'll just make it more vague and say "loaded by (but not updated by) Dagster".

Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/source_asset.py:27]: sounds good

File: python_modules/dagster/dagster/core/execution/context/system.py (hunk 3 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/core/execution/context/system.py:505]: agreed, right now it's kinda an olive branch for people that are used to writing non-asset specific io managers, but I don't think it's a good idea to keep around forever.

**Review by sryza — APPROVED** [review]
ship it!

**Comment by OwenKephart** [comment]
for future reference, the snapshot diff looks a lot worse than it actually is. There are so many repeated lines in the file that git gets confused when one block gets deleted (it kinda thinks that each block turned into the next one down the line). Doing `git diff --diff-algorithm=histogram --color-moved=dimmed-zebra` gives a somewhat more accurate picture of the actual changes.

From there, I basically manually combed through the output to make sure that there weren't any unexpected changes. The big difference (other than keys being changed) are a) root_manager ResourceDefSnaps are being removed and b) some op config fields no longer have an `inputs` entry. Both of these are expected as a result of the changes in this PR.

**Comment by OwenKephart** [comment]
for future reference, the snapshot diff looks a lot worse than it actually is. There are so many repeated lines in the file that git gets confused when one block gets deleted (it kinda thinks that each block turned into the next one down the line). Doing `git diff --diff-algorithm=histogram --color-moved=dimmed-zebra` gives a somewhat more accurate picture of the actual changes.

From there, I basically manually combed through the output to make sure that there weren't any unexpected changes. The big difference (other than keys being changed) are a) root_manager ResourceDefSnaps are being removed and b) some op config fields no longer have an `inputs` entry. Both of these are expected as a result of the changes in this PR.

**Pull request merged**